### ISSUE TITLE
FreeBSD implementation 

### ISSUE THREAD

**Issue by kris-nova** [kris-nova/kubicorn#253]
You heard me.

I want `kubicorn` to support running `kubernetes` on FreeBSD. It's my favorite operating system, and I will be damned if I own a project that doesn't work will with Beastie. 

I can code this as docker on FreeBSD is lots of fun, but if anyone else wants to get this going feel free to! 

![image](https://user-images.githubusercontent.com/13757818/29200308-7e5c0bec-7e11-11e7-9286-a2009424068e.png)

**Comment by bartekrutkowski** [comment]
So, where do I start? ;)

**Comment by xmudrii** [comment]
@bartekrutkowski Hello and welcome to kubicorn project. :)

As for DigitalOcean, I think best places to start are pull requests #240 and #217 
Bootstrap script are used to setup servers (in DO case Droplets) and they are located here https://github.com/kris-nova/kubicorn/tree/master/bootstrap
If you need any help, feel free to ping me here on kubicorn slack.

For Amazon, I don't use it so I can't help a lot about it. I believe @kris-nova should be able to do so. :)
Anyways, process should be very similar.

**Comment by bartekrutkowski** [comment]
@xmudrii I'm not sure I understand you - my question was particularly about the issue in question, that is, FreeBSD support :)

**Comment by xmudrii** [comment]
Sorry, my bad! I thought this is about supporting FreeBSD on cloud and not getting kubicorn working with it. :sweat: 

Therefore, @kris-nova will need to help you with this one. :)

**Comment by kris-nova** [comment]
@bartekrutkowski here are the loose steps I had in mind to get this working

1. Pick a cloud 
    a. AWS has a FreeBSD AMI in the Marketplace I have used before - it works great
    b. DigitalOcean has full FreeBSD support (I would start here)
2. Create a profile
    a. Look in the `/profiles` directory for example (I would just copy the existing DO one and start tweaking)
3. Map the profile to a flag
    a. Look in the `cmd/create.go` file for example
4. Deploy a cluster
5. See what breaks
6. Fix it
    a. I know this is going to be hard as we will have to write a bash script to pull a FreeBSD `kubeadm` binary manually (Ping me if we need to host it ourselves)
    b. And also figure out how we are going to handle `systemd` :disappointed: 
    c. Get a cluster up and running and save your work 
7. Pull request

**Comment by alexellis** [comment]
Hi nice to meet you. This sounds like a very cool thing to do. How good is the Docker/container runtime support on FreeBSD? (I haven't tried it)

The folks at Packet also have Official OS Support for FreeBSD on their [servers](https://www.packet.net/bare-metal/servers/type-0/) I read [this article](https://thenewstack.io/cncf-packet-team-provide-free-infrastructure-cloud-developers/) on the New Stack and it suggests there may free a way to get free server credits for open source projects.

**Comment by prateekgogia** [comment]
I don't think Docker has official support for FreeBSD, but I could be wrong too. 
If packet gives credits for opensource projects, may be we should add support for packet sooner and leverage that.

**Comment by kris-nova** [comment]
It's "experimental" https://wiki.freebsd.org/Docker

Which in Kubernetes land means it's ready to go

**Comment by kris-nova** [comment]
...the only way we are ever gonna get there is if we start pushing the boundaries

**Comment by freebsdfrau** [comment]
Hello. I am a FreeBSD src and ports committer. You can blame me for the boot loader menu, large swaths of the installer, bsdconfig, sysrc, libdpv, dpv, libfigpar, VNET jail examples, DTrace utilities, and many more items as well as several ports including sysutils/cmocka and many others.

I am also a friend of @kris-nova -- she's awesome!

TL;DR: I'd like to help with this.

**Comment by bartekrutkowski** [comment]
Everyone, few comments:

- saying "experimental" about Docker based on the quoted wiki page is a _huge_ understatement - what that page was describing was a one persons' private attempt at Docker for FreeBSD, one that was never nowhere near being finished and that haven't seen a commit in over two years
- systemd on FreeBSD is not going to happen, it hooks too deep into Linux kernel to be remotely possible and/or sane
- I am not aware of any `kubeadm` binary for FreeBSD, where does it exist?

The part about writing a bash script for FreeBSD being a hard part made me realise I've fallen into a nicely engineered troll trap, good job :)

**Comment by kris-nova** [comment]
So this isn't a troll trap - i really want to do this - and I wasn't saying the bash would be the hard part I was saying that docker, and the systemd challenges were going to be hard...

As far as `kubeadm` on FreeBSD.. 

```
GOOS=freebsd 
GOARCH=amd64
```

ought to do the trick..

**Comment by kris-nova** [comment]
Also a note on tone @bartekrutkowski please make sure to keep these technical discussions postive and helpful. Welcome to the project. https://github.com/kris-nova/kubicorn/blob/master/CODE_OF_CONDUCT.md#our-standards

**Comment by kris-nova** [comment]
Goals for initial bootstrapping

 - Docker 1.11 (or greater) running in FreeBSD
 - Systemd alias layer
 - configure IPFW to filter on layer 2 as documented in the ipfw(8) manpage to replace ebtables
 - Ports
    - `net/socat`

**Comment by freebsdfrau** [comment]
Fellow FreeBSD developer Brad Davis (brd@) adds:

> you should talk to AllanJude about the docker stuff
> there was a BSDCam session about some stuff
> we need libjail(not started) and libifconfig (in progress by Marie-Helene(sp?)

**Comment by freebsdfrau** [comment]
> net/socat

Hmmm, didn't know about socat until today. This looks cool.

I was originally thinking netgraph or epair/if_bridge with VNET:

[https://svnweb.freebsd.org/base/head/share/examples/jails/]

But socat would remove the dependency on a custom [VNET enabled] kernel.

**Comment by kris-nova** [comment]
@freebsdfrau as much I love compiling the FreeBSD kernel - I think it would be best to lean on `socat` thoughts?

**Comment by freebsdfrau** [comment]
Agreed

**Comment by rtyler** [comment]
I have been making steady progress on bringing the older Docker daemon patches [up to speed with Docker v17.05.0-ce](https://github.com/moby/moby/compare/17.05.x...rtyler:current-17.05) which should enable Kubicorn to run on FreeBSD.

I'm hoping to [meet up with others](https://twitter.com/agentdero/status/897538605265993728) who are interested in this work on Sept. 5th (tentatively) in Berkeley (see [here](https://wiki.freebsd.org/DockerHackDay2017)).


Either way, how cool would K8s on FreeBSD in 2017 be.

**Comment by kris-nova** [comment]
I'll bring kubicorn shirts to the meetup to say thanks

**Comment by freebsdfrau** [comment]
I'll bring Smule shirts too ;D we can have a swag exchange, teehee. I'm a medium T-Shirt size (preferably in women's if they are non-unisex). Anyone else want to chime-in on T-Shirt sizes? If not I'll plan to bring a mix of sizes (our shirts are unisex)

**Comment by riggs-** [comment]
How did the Docker hack day go?

**Comment by GuillaumeBibaut** [comment]
Any news about this support? I'm using FreeBSD as my fav OS as well. Really interested in the kubernetes, and so in kubicorn. (This is not really helping, but I can't think another way to contact)
